### ISSUE TITLE
Adding custom filter in customer modules

### ISSUE THREAD

**Issue by jefenga** [opensourcepos/opensourcepos#2358]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 
- OSPOS git commit hash is: 
- PHP version is: 
- MySQL or MariaDB version is: 
- OS and version is: 
- WebServer is: 
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### Issue / Bug  / Question / New Feature

Please write your issue here. If a bug, please make sure to provide as much information as possible including configuration settings (e.g. Decimals set, Tax mode), language and steps to reproduce the bug.

**Comment by jefenga** [comment]
hello, I have multiple stock location and each customer is belong to a certain stock ( eg. customer juma is belong to stock A, he would not to be served on stock B, it because i follow the customer so i need to keep track of them)  I need to add filter for the customer module. like the item module, can any one enlighten me where should i start

**Comment by objecttothis** [comment]
@jefenga This is something that would be need to be added to the application.  I think the best way to do this would be to extend attributes to People and have an attribute called stock that is a multi-select DROPDOWN.  However it may be better to make something like this be a core aspect of the customer themselves rather than an attribute.  @daN4cat does this feel more like category in items to you (i.e., should be a core aspect of customer rather than an attribute)?

**Comment by daN4cat** [comment]
@objecttothis I thought about the best decision criteria between what's core and what's not.
This is what I have in mind as principles:
- Core is always required, essence of the entity we model
- Attribute is optional and contestualised to the application utilisation 

There is also an important trade off behind that distinction, core is a dimensional aspect and queries faster, attribute is a joined aspect and queries slower. I mean database wise.

If we approach the data/object modelling in a consistent way like the above we can implement a general pattern in a base class and reduce the code complexity/repetition.

With regards to code and switches, we need to leverage more object oriented design and polymorphism. Instead of adding conditionals all over the places we should specialise and instance the correct subclass, delegating to specialised object instances the execution of desired functionality.

**Comment by daN4cat** [comment]
With regards to the specific issue here, I struggle to understand the connection stock location / customer.
Stock location is an internal view of the system, customer should not be linked to where something is outside go to a shop.
Stock location is not shop location.

So what is the problem we are trying to solve here? What's the  business case?

**Comment by jefenga** [comment]
@daN4cat. The case here is the customers always buying from one location(ie in this case), so when i register a new customer   i need  to specified this customer is belong to certain location, so what i was asking is how can i link customer with stock location, And what would be the best approach to filter a customer with stock location.

I apologize for not being word-perfect in English

**Comment by objecttothis** [comment]
@dan4cat I agree regarding your definition of what is core and what is attribute. If we extended attributes to People (employees, suppliers, customers) then we should follow that definition.

**Comment by objecttothis** [comment]
@dan4cat regarding this issue, please clarify what you mean by stock location is not shop location? Do you mean that stock location was meant to dictate where within the shop the stock is found and not meant to dictate that the stock is located at a warehouse, shop#2 or shop#5?  If I understand the OP, they have multiple locations where they have items and only want to show customers in the customer view for that particular location the same way the items view only show items for a stock location.

**Comment by daN4cat** [comment]
@jefenga stock is a back room story, customers come to your shop.
I assume this is more for statistical / insight purposes not for other ones because customer are free to decide where to go.
It sounds like you are forcing a connection that doesn't exist.

One shop can have different stock locations: shelf, stock room, warehouse.

**Comment by objecttothis** [comment]
@daN4cat Is there a reason why Stock location can't be generic so that people with multi-site operations can't use it to declare which shop or warehouse an item is located?  I also thing we could benefit from more user documentation since I myself assumed that Stock location was exactly the way that @jefenga is using it.  Our local operation only has one shop but since we started out by using a custom field (and now an attribute) to show which section of the shop the item is located in, I never implemented Stock location in our shop.

The way stock location is currently written, I think it's technically already agnostic, other than the fact that if you want to do both in-store stock location and in-company stock location then you have to do like what we do and use an attribute for the in-store stock location and stock location dictates in-company stock location.  

The feature that @jefenga is requesting would need to be optional since it makes no sense if a user is currently using stock location as an in-store location (shelf, back room, etc).

I think it would just involve adding a relationship in the database with a stock_id and an optional field in the customer view and then allowing a filter in the customers table by stock location.

**Comment by daN4cat** [comment]
You could hard link stock location with shop, but personally I would not model it like that. I still see a distinction between the two and offers more flexibility.
One warehouse (stock location) could serve one or more shops.
A shop stock room is a stock location were stock could be moved to another shop.
In receiving there is a requisition transaction to cover the case where you move stock from one place to another and get a receipt.
@jekkos can cover more this topic.

Multi shop, multi point of sale, multi stock locations.... multi tenant, SaaS. We are aiming at enterprise solutions here :-)

**Comment by objecttothis** [comment]
@daN4cat A few questions:
A distinction between in-store stock location and in-organization stock location (for lack of a better term) makes sense to me because you still may want the flexibility of tracking both.  Are you opposed to adding the ability then to track in-organization stock location?

There are many features to a full-fledged enterprise solution that I myself have no interest in implementing, but I do have interest in the ability of one install being accessed by multiple shop locations and being able to see each other's inventory to better facilitate moving stock from one shop location to another.  In essence the functionality of "shop location" (at a basic) level would be near identical to what we currently have in place with stock location except that there would be the added ability to associate employees with a shop location so that defaults can be set to show items associated with that shop location and for customers to be associated with a shop location which would then allow filters show you customers for one shop location or another.

**Comment by daN4cat** [comment]
We should reason around the shop model first and the dependencies because there are many before the stock story.

**Comment by objecttothis** [comment]
@daN4cat agreed.  How do you propose we proceed with that?

**Comment by tru247** [comment]
Good evening Opensourcepos/Opensourcepos

I propose to add mobile payments. Starting with mpesa. What is the
onboarding process?

**Comment by daN4cat** [comment]
@tru247 off topic, please create a new issue for that. Thanks

**Comment by jekkos** [comment]
@objecttothis I have a docker container on docker hub with shop multilocation support, with functionality you describe here.

**Comment by WebShells** [comment]
We should keep in mind different needs for creating multisite ready branch, receipt data: shop address, phone number, return policy, its own default discount, etc... Dont you think so?
So the data should be retreived based on which stock location the user is assigned to, or selected upon login... 
Cashup, gift cards, clients based on selected store as well :/
